packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+  /// retrieved from [toPrimitives] is removed from the restoration data. When
+  /// it changes from false to true, [toPrimitives] is invoked to add the latest
+  /// restoration information provided by this property to the restoration data.
+  bool get enabled => true;
+
+  bool _disposed = false;
+
+  @override
+  void dispose() {
+    assert(_debugAssertNotDisposed());
+    _owner?._unregister(this);
+    super.dispose();
+    _disposed = true;
+  }
+
+  // ID under which the property has been registered with the RestorationMixin.
+  RestorationId _id;
+  RestorationMixin _owner;
+  void _register(RestorationId id, RestorationMixin owner) {
+    assert(_debugAssertNotDisposed());
+    assert(id != null);
+    assert(owner != null);
+    _id = id;
+    _owner = owner;
+  }
+  void _unregister() {
+    assert(_debugAssertNotDisposed());
+    _id = null;
+    _owner = null;
+  }
+
+  /// The [State] object that this property is registered with.
+  @protected
+  State get state {
+    assert(_debugAssertNotDisposed());
+    return _owner;
+  }
+
+  /// Whether this property is currently registered with a [RestorationMixin].
+  @protected
+  bool get isRegistered {
+    assert(_debugAssertNotDisposed());
+    return _id != null;
+  }
+
+  bool _debugAssertNotDisposed() {
+    assert(() {
+      if (_disposed) {
+        throw FlutterError(
+          'A $runtimeType was used after being disposed.\n'
+          'Once you have called dispose() on a $runtimeType, it can no longer be used.'
+        );
+      }
+      return true;
+    }());
+    return true;
+  }
+}
+
+/// Manages the restoration data for a [State] object of a [StatefulWidget].
+///
+/// Restoration data can be serialized out and at a later point in time be used
+/// to restore the stateful members in the [State] object to the same values
+/// they had when the data was generated.
+///
+/// This mixin organizes the restoration data of a [State] object in
+/// [RestorableProperty]. All the information that the [State] object wants to
+/// get restored during state restoration need to be saved in a subclass of
+/// [RestorableProperty]. For example, to restore the count value in a counter
+/// app, that value should be stored in a member variable of type
+/// [RestorableInt] instead of a plain member variable of type [int].
+///
+/// The mixin ensures that the current values of the [RestorableProperty]s are
+/// serialized out as part of the restoration state. It is up to the [State] to
+/// ensure that the data stored in the properties is always up to date. When the
+/// widget is restored from previously generated restoration data, the values of
+/// the [RestorableProperty]s are automatically restored to the values that had
+/// when the restoration data was serialized out.
+///
+/// Within a [State] that uses this mixin, [RestorableProperty]s are usually
+/// instantiated to initialize member variables or within [State.initState].
```

Comment:
```suggestion
/// instantiated to initialize member variables within [State.initState].
```
Is the "or" not needed?  If you did mean to include the "or", then this sentence is a little awkward, and I'm not sure what you mean.